x-pack/filebeat/input/http_endpoint/input.go

```
@@ -0,0 +1,290 @@
+	eventObject  *map[string]interface{} // Current event object
+}
+
+// NewInput creates a new httpjson input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	// Build outlet for events.
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
+		Processing: beat.ProcessingConfig{
+			DynamicFields: inputContext.DynamicFields,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &HttpEndpoint{
+		config:       conf,
+		log:          logp.NewLogger(inputName),
+		outlet:       out,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+	}
+
+	// Create an instance of the HTTP server with the beat context
+	in.server, err = createServer(in)
+	if err != nil {
+		return nil, err
+	}
+
+	in.log.Infof("Initialized %v input on %v:%v", inputName, in.config.ListenAddress, in.config.ListenPort)
+	return in, nil
+}
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *HttpEndpoint) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go in.run()
+		in.workerWg.Done()
+	})
+}
+
+func (in *HttpEndpoint) run() {
+	defer in.log.Infof("%v worker has stopped.", inputName)
+	err := in.server.Start()
```

You can ignore this error, as it is always nil.
```
func (in *HttpEndpoint) run() {
	defer in.log.Infof("%v worker has stopped.", inputName)
	in.server.Start()
}
```

---

Also, I have a feeling this `run` returns immediately, as it is starting a goroutine. It is probably not necessary to use a goroutine there. I am not 100% sure as I haven't run the code.

---

If func (h *HttpServer) Start() is not a goroutine then the http server does not close when doing CTRL+C, and it gets stuck shutting down. But I removed the return for now.